Add aspect ratio presets support via theme.json

jffng opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR adds support for aspect ratio presets via theme.json. It adds by default the same presets that are available for the crop utilities in the image block.

## Why?
Part of https://github.com/WordPress/gutenberg/issues/38990

## How?
Bootstraps the presets via the theme.json class.

## Testing Instructions
- Check out this PR and verify you see the presets in the following places:
    - Core Image Block Inspector Controls
    - Core Image Block Croping tools (Toolbar)
    - Core Featured Image Inspector Controls
    - Core Cover block Inspector Controls
- Add your own presets via the theme (`settings.dimensions.aspectRatios`) and verify they appear
- Disable the default presets `settings.dimensions.defaultAspectRatios` -> `false` and verify they do not appear

<details>
<summary> Sample `theme.json`</summary>

```json
{
	"version": 3,
	"settings": {
		"dimensions": {
			"defaultAspectRatios": true,
			"aspectRatios": [
				{
					"name": "Extra Wide - 2:1",
					"ratio": "2/1",
					"slug": "2-1"
				}
			]
		}
	}
}

```

</details>

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A, backend only change. 

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/20684594/9050f661-1452-4699-8822-93b2fd52466c

jffng committed:
Remove defaultAspectRatio false.

mtias commented:
Could we just go with `ratio` and avoid `name` and `slug`?

jffng commented:
@mtias I believe `slug` is currently [required by the Style Engine](https://github.com/WordPress/gutenberg/blob/trunk/packages/style-engine/class-wp-style-engine.php#L33) in order to generate the preset. 

We can drop `name` if that's not helpful, but I was not sure if there are cases where we'd want to name a ratio differently, e.g. `Square` for `1/1`.

Mamaduka commented:
Hi, @jffng

I just wanted to check the status of the feature. Are we still aiming for WP 6.2?

ajlende commented:
I have some bandwidth to work on this right now if @jffng is busy with other things.

Commit:
Rename 1:1 to square

Commit:
Rename aspectRatio to aspectRatios

Commit:
Add defaultAspectRatios

Commit:
Add defaultAspectRatios setting

Commit:
Add dimensions to settingPropertiesComplete

Commit:
Add dimensions.aspectRatios to PATHS_WITH_MERGE

Commit:
Add theme presets to aspect ratio dropdown

Commit:
Configure defaultAspectRatios for aspect ratio dropdown

Commit:
Add support for aspect ratios.

ajlende commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -114,9 +114,19 @@ class WP_Theme_JSON_Gutenberg {
 	 * @var array
 	 */
 	const PRESETS_METADATA = array(
+		array(
+			'path'              => array( 'dimensions', 'aspectRatios' ),
+			'prevent_override'  => array( 'dimensions', 'defaultAspectRatios' ),
+			'use_default_names' => false,
+			'value_key'         => 'ratio',
+			'css_vars'          => '--wp--preset--aspect-ratio--$slug',
+			'classes'           => array(),
+			'properties'        => array( 'aspect-ratio' ),
+		),
```
I wanted to add just the settings for the image block aspect ratio tool in this PR, but it seems that the UI is coupled to the preset generation. So, even though this generated CSS isn't used, it had to be added to get the UI to work.

I don't think it's too big of a deal since we would add those presets for the `aspect-ratio` property anyway for the featured image in a follow-up, but it certainly surprised me that the code was coupled this way.

ajlende commented on lib/theme.json:
```diff
@@ -186,6 +186,51 @@
 		},
+		"dimensions": {
+			"defaultAspectRatios": true,
+			"aspectRatios": [
```
These default presets are currently unused as well so the UI can continue to be grouped by portrait and landscape to match the design. It's unfortunate that [we aren't using JSON with comments](https://github.com/WordPress/gutenberg/issues/35099), so I can't really comment in the code here about it.

ajlende replied:
I think I'm going to update the UI to use these and add a special case for the defaults to categorize them as landscape or portrait.

We could also categorize the theme supplied ones as well if we update the designs. Or we could choose to not categorize the defaults to be consistent with how other settings presets work. @richtabor maybe you have thoughts?

richtabor replied:
> Or we could choose to not categorize the defaults to be consistent with how other settings presets work 

You mean keep the "Theme" section, with the theme's defaults, v. putting those defaults within the proper landscape/portrait sections?

ajlende replied:
The theme ones would be as-is and the landscape and portrait get grouped into just default. For example, in the color/gradient presets, they are grouped by default and theme. And in global styles, there's an additional custom section.

![global styles color presets](https://user-images.githubusercontent.com/5129775/216399918-2cf36a43-c6a9-4629-9121-c55709076216.png)

richtabor replied:
> The theme ones would be as-is and the landscape and portrait get grouped into just default. For example, in the color/gradient presets, they are grouped by default and theme. 

Would we loose the separation of portrait and landscape? 

Is it a heavy lift to get theme presets in? If so, I'm ok dropping support for theme-specific presets (for now) and circle back around for that.

ajlende replied:
I can organize it however you'd like. I pushed up a change to include the defaults from the default theme.json categorized under landscape and portrait already. It also isn't any more difficult to change how they're categorized if you think it makes sense to do it differently.

richtabor replied:
Ok cool, good deal.

Commit:
Update docs

Commit:
Add comment about two-value aspect-ratio CSS

Commit:
Add missing 2:3 aspect ratio

Commit:
Refactor aspect ratios to come from theme.json

Commit:
Move aspect-ratio-dropdown into components

Commit:
Rename AspectGroup to AspectRatioGroup

ajlende commented:
I've been working on creating a new select component that allows for dividers and headings to match the designs. I can't reuse the existing dropdown because it's a toolbar dropdown, not an input/select dropdown. And I can't use the existing SelectControl or CustomSelectControl because they don't support custom children like our section headers or dividers.

I figured I could reuse the existing presentational components of the CustomSelectControl, but they come with [too much baggage](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/custom-select-control/index.js): backwards compatibility, future compatibility, experimental options and features that the new component wouldn't use. It also seems bad to not use those presentational components since they should, in theory, be the right thing to use.

So in the interest of getting this shipped and iterated on, I'm going to deviate from the designs and use the existing SelectControl with the "current aspect" option at top followed by the theme options and then the default options without any of the separators or headers.

richtabor commented:
> I'm going to deviate from the designs and use the existing SelectControl with the "current aspect" option at top followed by the theme options and then the default options without any of the separators or headers.

That works. We can iterate towards it. 

>  followed by the theme options 

After talking with @mtias, it would better to hold off on the theme defaults aspect of this — to give us time to ensure the theme aspect ratios preset spec is clear/expected/well-defined.

ndiego commented:
Hey @richtabor @ajlende, this PR is on the [6.3 project board](https://github.com/orgs/WordPress/projects/103/views/1). It's been a while since this was worked on. Is this still something we are aiming to get into 6.3?

richtabor commented:
> Is this still something we are aiming to get into 6.3?

Not not currently; we can remove.

unscripted commented:
I would love to see this in an upcoming release. Unless I'm missing it, we cannot add a custom aspect ratio via theme.json.

strarsis commented:
+1, this would be such a great addition! Being able to define the crop interactively for each image without extra plugins.

fabiankaegy commented:
I just rebased this PR in order to make it easier to test and evolve the API :)

fabiankaegy committed:
run docs build

fabiankaegy committed:
fix use theme.json defined aspect ratio options in aspect ratio tools panel

fabiankaegy committed:
fix casing of aspect-ratio custom property

fabiankaegy committed:
fix remove unwanted formatting change

fabiankaegy committed:
fix crop aspect ratio tools

fabiankaegy committed:
fix tests by passing in the available aspect ratio options

This was failing before because in the test case here the theme.json options are not loaded

fabiankaegy committed:
fix label in test case

Mamaduka commented on packages/block-editor/src/components/aspect-ratio-dropdown/index.js:
```diff
@@ -0,0 +1,150 @@
+	const { isInProgress, aspect, setAspect, defaultAspect } =
+		useImageEditingContext();
+
+	const [ defaultRatios ] = useSettings( 'dimensions.aspectRatios.default' );
+	const [ themeRatios ] = useSettings( 'dimensions.aspectRatios.theme' );
+	const [ showDefaultRatios ] = useSettings(
+		'dimensions.defaultAspectRatios'
+	);
```
Let's combine the setting getters - `useSettings( 'dimensions.aspectRatios.default', 'dimensions.aspectRatios.theme', 'dimensions.defaultAspectRatios' )`

Related: https://make.wordpress.org/core/2024/03/09/miscellaneous-editor-changes-in-wordpress-6-5/#new-usesettings-hook-for-reading-block-instance-settings

fabiankaegy replied:
Ahh very good call 🤦

fabiankaegy replied:
Fixed in https://github.com/WordPress/gutenberg/pull/47271/commits/001391b1a09287d798116a7e9fa1b29c05e4dffe

Mamaduka commented on packages/block-editor/src/components/aspect-ratio-dropdown/index.js:
```diff
@@ -0,0 +1,150 @@
+	);
+}
+
+function ratioToNumber( str ) {
+	// TODO: support two-value aspect ratio?
+	// https://css-tricks.com/almanac/properties/a/aspect-ratio/#aa-it-can-take-two-values
+	const [ a, b, ...rest ] = str.split( '/' ).map( Number );
+	if (
+		a <= 0 ||
+		b <= 0 ||
+		Number.isNaN( a ) ||
+		Number.isNaN( b ) ||
+		rest.length
+	) {
+		return NaN;
+	}
+	return b ? a / b : a;
+}
+
+function presetRatioAsNumber( { ratio, ...rest } ) {
+	return {
+		ratio: ratioToNumber( ratio ),
+		...rest,
+	};
+}
```
Having unit tests for these utils would be helpful.

fabiankaegy replied:
Added a test in https://github.com/WordPress/gutenberg/pull/47271/commits/18c204517ea7f6f82b416f2e56f00355216fcbb8

fabiankaegy committed:
fix combine multiple useSettings call into one

fabiankaegy committed:
add test case for ratioToNumber function

Mamaduka commented on packages/block-editor/src/components/image-editor/index.js:
```diff
@@ -6,11 +6,11 @@ import { ToolbarGroup, ToolbarItem } from '@wordpress/components';
 /**
  * Internal dependencies
  */
+import AspectRatioDropdown from '../aspect-ratio-dropdown';
```
What was the reason for extracting this in a separate component? It seems to be used only by the image editor.

fabiankaegy replied:
That is a very good question. This was already done before I took over the branch. 🤔

Mamaduka replied:
One reason I think we should group it with the Image Editor is that the dropdown also resizes the image on the server. Having it as a stand-alone component could be misleading.

fabiankaegy replied:
That's a great reason to keep it together :) moved it in https://github.com/WordPress/gutenberg/pull/47271/commits/8a3088c6f56223871f1c8c3dd193791ad4f48944

Mamaduka commented on packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js:
```diff
@@ -96,14 +35,49 @@ export default function AspectRatioTool( {
+		},
+		...( showDefaultRatios ? defaultOptions : [] ),
+		...( themeOptions ? themeOptions : [] ),
+		...( options ? options : [] ),
```
Unlike `trunk`, this will merge provided options instead of overriding the defaults.

fabiankaegy replied:
Yes and no. The defaults are now no longer hardcoded here in this JS file but instead are now coming from the core `theme.json` file. 

So the `showDefaultRatios` check determines if they display or not. The `options` currently were not used in the Gutenberg codebase at all. I now needed to use them in the unit tests since the global styles are not populated there. But I'm not sure this is breaking the expected behavior 🤔

fabiankaegy replied:
Fixed in https://github.com/WordPress/gutenberg/pull/47271/commits/7f6223fe1c8f24eeed53b9e83a5697006201587e

Mamaduka commented on packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js:
```diff
@@ -117,7 +91,7 @@ export default function AspectRatioTool( {
 				label={ __( 'Aspect ratio' ) }
 				value={ displayValue }
-				options={ options }
+				options={ aspectRatioOptions }
```
Instead of merging `options` above, we can pass `options ?? aspectRatioOptions` and preserve the previous component behavior.

fabiankaegy replied:
That works :)

fabiankaegy replied:
Fixed in https://github.com/WordPress/gutenberg/pull/47271/commits/7f6223fe1c8f24eeed53b9e83a5697006201587e :)

fabiankaegy committed:
fix restore previous behavior for passing manual options

fabiankaegy committed:
Move AspectRatioDropdown to Image Editor components

Mamaduka commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -397,8 +407,10 @@ class WP_Theme_JSON_Gutenberg {
 		'dimensions'                    => array(
-			'aspectRatio' => null,
-			'minHeight'   => null,
+			'aspectRatio'         => null,
```
Unrelated: I noticed that you can't disable aspect ration for Image block using this setting. cc @draganescu

draganescu replied:
Was it supposed to be 😄 What is the expected behaviour here:

- does setting dimensions ≻ aspect ratio ≻ null disable the aspect ratio tools somewhere else?
- should it disappear from the UI of the image block? It does not appear in styles, but in settings, whereas for the cover block the UI appears in styles not in settings. What's preffered?
- should it be null or false? When I set theme.json ≻ settings ≻ dimensions ≻ aspectRatio the editor says it should be bool

Mamaduka replied:
The initial `null` value is how theme json defines flag settings, and it's not important at the moment.

Setting it to `false` disables the aspect ratio UI for blocks. You can test it with the Cover block. I think behavior should be consistent across the blocks that use aspect ratio control.

cc @andrewserong

draganescu replied:
Yes, indeed setting to false hides this dimension tool from the cover block but it does not hide it from the image block. The reason is that the cover block uses the supports while the image block plops the control in the inspector.

Mamaduka replied:
Yes. I'm aware of that. Still, the behavior should be consistent.

@fabiankaegy, do you mind creating an issue for the Image and Featured Image block follow-ups we discussed in Slack?

draganescu replied:
> Yes. I'm aware of that.

I knew that, you told me where to look! 😂  I mentioned here for archaeology reasons.

fabiankaegy replied:
Created a follow up issue for the aspect ratio block support cleanup: https://github.com/WordPress/gutenberg/issues/61432

fabiankaegy committed:
fix use aspect ratio options from theme.json in post featured image block

fabiankaegy commented:
@ajlende wanted to tag you specifically here for feedback since @oandregal mentioned you recently worked on this area and also you were one of the earlier contributors to this PR before I took it over :)

bacoords commented:
Did some user testing on this branch and was able to get through all of the instructions testing custom & default aspect ratios across multiple block themes.

bacoords commented:
One note about the testing instructions. Should it list the specific blocks that use aspect ratios? Here's what I tested but I may have missed some: 

- Image 
- Cover
- Featured Image

fabiankaegy commented:
> One note about the testing instructions. Should it list the specific blocks that use aspect ratios? Here's what I tested but I may have missed some:
> 
> * Image
> * Cover
> * Featured Image

@bacoords I'll update the testing instructions to be a little more concise. But you are correct those are all the instances. The one thing to note it that for the image block it shows in two places. The sidebar setting for the aspect ratio and then also in the croping tools in the sidebar

Mamaduka reviewed (approved):
Let's merge this. I don't see any blockers here; we can handle feature unification in follow-ups.

fabiankaegy commented:
I just created a sync PR on WordPress Develop for the PHP updates to the theme.json class https://core.trac.wordpress.org/ticket/61137
https://github.com/WordPress/wordpress-develop/pull/6567

swissspidy commented:
Internationalization needs to be fixed before this lands in core. See #61767

fabiankaegy commented:
@swissspidy thanks for calling this out. I have created #61774 to fix this here in Gutenberg and also updated my PR in core to include the fix. 

Question for you. These strings were previously already present in core. Just hardcoded in JS and piped through `__()`. Will these existing translations get mapped to the new strings automatically or will they need to get translated from scratch?

swissspidy commented:
They will be marked as fuzzy because now they have proper context (_x vs __)

fabiankaegy commented:
## Dev Note Draft 


# Add custom Aspect Ratio presets through `theme.json`

WordPress 6.6 adds a new two new properties to the `settings.dimensions` object in `theme.json`. `settings.dimensions.aspectRatios` allows defining your own custom Aspect Ratio presets. These presets will be used by any block that uses the `aspectRatio` block support. In core that means the Image, Featured Image, and Cover block. 

```json
{
	"version": 2,
	"settings": {
		"dimensions": {
			"aspectRatios": [
				{
 					"name": "Square - 1:1",
 					"slug": "square",
 					"ratio": "1"
 				},
 				{
 					"name": "Standard - 4:3",
 					"slug": "4-3",
 					"ratio": "4/3"
 				},
			]
		}
	}
}
```

Additionally the `settings.dimensions.defaultAspectRatios` key allows you to disable the list of core aspect ratio presets.

```json
{
	"version": 2,
	"settings": {
		"dimensions": {
			"defaultAspectRatios": false
		}
	}
}
```
